### DISCUSSION TITLE
Pass raw YUV 4:2:0 Data to JpegEncoder

### DISCUSSION THREAD

**Discussion by joedemax** [SixLabors/ImageSharp#2495] [Q&A]
Hi,

I have been trying to find the very fastest way to convert frames I am rendering to the GPU to a compressed format. Jpeg with YCbCrRatio420 seems to be the fastest so far, but I can't quite find a way to pass it raw 4:2:0 data to work on. I am already converting these frames to 4:2:0 on the GPU (partially to enable faster readback) and as a result I already have them in this format.   However, I can't find a way to pass my raw 4:2:0 data to the JpegEncoder or an Image, instead I have to convert back to RGB first before ImageSharp presumably flips it back to 4:2:0. 

This seems immensly wasteful. Is there anyway around this?

**Comment by antonfirsov** [comment]
This would require cleaning up and exposing APIs which are currently implementation details of JpegEncoder which is a lot of work. It's unlikely we will ever do it, unless this turns out to be a popular feature request for some reason.

I'm not sure if there are any high level imaging libs out there exposing this level of control. If you are comfortable coding part of your logic in C/C++, you can most likely do it by using libjpeg-turbo directly.

**Reply by br3aker** [reply]
I don't think it's _that_ hard but too specific for a general purpose library like ImageSharp is.
Idea of exposing entire encoder as some sort of an interface may be a good solution for this. But again, jpeg-only encoder exposure may look strange compared to other formats IMO.

**Reply by JimBobSquarePants** [reply]
I don’t think it’s wise to expose so much inner workings. It’d be better if the @joedemax just copied what they needed from the source.
